### PR TITLE
fix Issue 5303 "Build fail with gcc 9"

### PR DESCRIPTION
**Pull request by biocodz**
Please refer to [Issue 5303 Build fail with gcc 9](https://github.com/facebook/rocksdb/issues/5303)
3 files are modified:
- db/internal_stats.h - added copy-assignment operator in `struct CompactionStats`
- db/version_edit.h - added copy-constructor in `struct FileDescriptor`
- utilities/persistent_cache/persistent_cache_util.h - changed `return std::move(t)` to `return t` to avoid the `copy elision` error.

Implementing copy-constructor and copy-assignment in pairs is a good practice recommended by C++ standard anyway. It is just enforced in GCC 9.1 (7.x allows it)

The build was tested on Ubuntu 18.04 and Windows with VS 2017

The changes are legit and good also for earlier less strict compilers supporting C++11

### CODE REVIEW

**Comment by siying** [comment]
I can merge it if tests pass, but we are not mechanism to guarantee it doesn't break again in the future. Any thought?

**Comment by biocodz** [comment]
Well, there is never a guarantee something won't break in the future. If it breaks we'll fix it again. 
These changes only introduced some missing code and fixed a move misuse. They didn't introduce any new logic. So, this particular fix should be stable.

File: db/internal_stats.h [hunk 1]
```diff
@@ -237,6 +237,27 @@ class InternalStats {
       }
     }
 
+	CompactionStats & operator=(const CompactionStats& c) {
```
Comment (siying) [inline db/internal_stats.h]: We use space only for indenting, and there is tab here, which is not allowed. Please fix it.

Comment (biocodz) [reply db/internal_stats.h]: Fixed. Travis job glitched and needs to be re-run.

**Comment by biocodz** [comment]
Again this is a Travis glitch crashing the `TEST_GROUP=platform_dependent` (nothing to do with this fix):
```
apt-get install failed
The command "sudo -E apt-get -yq --no-install-suggests --no-install-recommends $(travis_apt_get_options) install curl g++-8 libbz2-dev libgflags-dev libsnappy-dev mingw-w64 zlib1g-dev" failed and exited with 100.
```

**Comment by biocodz** [comment]
The last failure has nothing to do with this fix:
```
db_test::DBTest.SparseMerge State: Completed
Note: Google Test filter = DBTest.SparseMerge
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from DBTest
[ RUN      ] DBTest.SparseMerge
c:\projects\rocksdb\db\db_test.cc(1233): error: Expected: (dbfull()->TEST_MaxNextLevelOverlappingBytes(handles_[1])) <= (20 * 1048576), actual: 32772361 vs 20971520
[  FAILED  ] DBTest.SparseMerge (6253 ms)
[----------] 1 test from DBTest (6253 ms total)
[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (6254 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] DBTest.SparseMerge
 1 FAILED TEST
```

**Comment by iosmanthus** [comment]
@siying In `ArchLinux`, the `GCC` has been upgraded to `9.1.0` 😰. It is awkward to specify the `CXXFLAGS`  manually to bypass such error. It will be really great if this PR merged.

**Comment by biocodz** [comment]
A test irrelevant to this fix failed
```
env_test::DefaultEnvWithoutDirectIO/EnvPosixTestWithParam.DecreaseNumBgThreads/0 State: Completed
Note: Google Test filter = DefaultEnvWithoutDirectIO/EnvPosixTestWithParam.DecreaseNumBgThreads/0
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from DefaultEnvWithoutDirectIO/EnvPosixTestWithParam
[ RUN      ] DefaultEnvWithoutDirectIO/EnvPosixTestWithParam.DecreaseNumBgThreads/0
c:\projects\rocksdb\env\env_test.cc(620): error: Value of: env_->GetThreadPoolQueueLen(Env::Priority::HIGH)
  Actual: 1
Expected: (unsigned int)0
Which is: 0
```

**Comment by biocodz** [comment]
Travis error is unrelated to this PR:
```
CMake Error at java/CMakeLists.txt:324 (message):
  Failed downloading C:/projects/rocksdb/java/test-libs/junit-4.12.jar
-- Configuring incomplete, errors occurred!
```

**Comment by biocodz** [comment]
Closing this pull request as it was fixed in [request 5553](https://github.com/facebook/rocksdb/pull/5553)

**Pull request closed without merging**